statistics_service/native/src/impl/StatisticsService.cpp

```diff
@@ -1,164 +1,234 @@
+                        break;
+                    default:
+                        ALOGE("%s:%d: Error when pushing values", __FUNCTION__, __LINE__);
                         break;
-                    }
                 }
+                ++current_value_index;
             }
-            return values;
         }
+    } else {
+        ALOGE("%s:%d: std::cout << Error reading file ", __FUNCTION__, __LINE__);
+    }
+}
 
-        float StatisticsService::getValue(CallbacksId id)
-        {
-            std::vector<float> read_values = readTemperatures();
-
-            switch (id) {
-                CallbacksId::CpuTemperature:
-                    return read_values[0];
-                CallbacksId::GpuTemperature:
-                    return read_values[1];
-                CallbacksId::AmbientTemperature:
-                    return read_values[2];
-                CallbacksId::AverageCpuTemperature:
-                    return read_values[0];
-                CallbacksId::AverageGpuTemperature:
-                    return read_values[1];
-                CallbacksId::AverageAmbientTemperature:
-                    return read_values[2];
-                CallbacksId::MaxCpuTemperature:
-                    return read_values[0];
-                CallbacksId::MaxGpuTemperature:
-                    return read_values[1];
-                CallbacksId::MaxAmbientTemperature:
-                    return read_values[2];
-                default:
-                    return 0.0;
-            }
-        }
+float StatisticsService::GetCpuValue() {
+    float cpu_value = *(all_cpu_temperatures_.begin() + iterator_all_cpu_temps_);
+    ++iterator_all_cpu_temps_;
+    iterated_cpu_values_.push_back(cpu_value);
+
+    if(iterator_all_cpu_temps_ == all_cpu_temperatures_.size())
+        iterator_all_cpu_temps_ = 0;
+    
+    return cpu_value;
+}
+
+float StatisticsService::GetGpuValue() {
+    float gpu_value = *(all_gpu_temperatures_.begin() + iterator_all_gpu_temps_);
+    ++iterator_all_gpu_temps_;
+    iterated_gpu_values_.push_back(gpu_value);
+
+    if(iterator_all_gpu_temps_ == all_gpu_temperatures_.size())
+        iterator_all_gpu_temps_ = 0;
+
+    return gpu_value;
+}
+
+float StatisticsService::GetAmbientValue() {
+    float ambient_value = *(all_ambient_temperatures_.begin() + iterator_all_ambient_temps_);
+    ++iterator_all_ambient_temps_;
+    iterated_ambient_values_.push_back(ambient_value);
+
+    if(iterator_all_ambient_temps_ == all_ambient_temperatures_.size())
+        iterator_all_ambient_temps_ = 0;
+
+    return ambient_value;
+}
+
+float StatisticsService::CalculateAverageCpu() {
+    if(iterator_all_cpu_temps_ == 1)  //when we start over, vector values have to be cleared excepting first element
+        iterated_cpu_values_.erase(iterated_cpu_values_.begin() + 1, iterated_cpu_values_.end()); 
+
+    float sum = std::accumulate(iterated_cpu_values_.begin(), iterated_cpu_values_.end(), 0.0f);

```

Comment:
TIL: `std::reduce` poate fi mai rapid: 
https://en.cppreference.com/w/cpp/algorithm/reduce
https://stackoverflow.com/questions/47144083/why-has-stdreduce-been-added-in-c17